### PR TITLE
Add pre boot tuning operations

### PR DESCRIPTION
**Pull request by yanirq**
Very early tuning:  This is needed because a configuration needs to be set before early-boot things start. 
The two configurations we change are (A) systemd CPUAffinity and (B) kernel workqueue cpumask, we do that using a custom initrd image we add as an overlay (set in boot-loader entries) 

This relies on setting a kernel argument : non_isolated_cpus that should hold a list of reserved cpus in the format of 1-2,5,9-10.

### CODE REVIEW

File: features/performance/assets/mc-pre-boot-worker-tuning.yaml.in [hunk 1]
```diff
@@ -0,0 +1,44 @@
+apiVersion: machineconfiguration.openshift.io/v1
+kind: MachineConfig
+metadata:
+    labels:
+        machineconfiguration.openshift.io/role: worker-rt
+    name: 11-rt-kernel-worker-rt
+spec:
+    kernelArguments:
+        - 'non_iso_cpumask=00000005'
```
Comment (vladikr) [inline features/performance/assets/mc-pre-boot-worker-tuning.yaml.in]: I think this needs to be calculated based on what we set in `isolcpus` or what does this mask mean?

File: features/performance/assets/pre-boot-tuning.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+set -euo pipefail
+
+# TODO: improve check for applied configuration
+if grep -q "CPUAffinity" "/etc/systemd/system.conf"; then
+    echo "Pre boot tuning configuration already applied"
+    # Move kernel rcuo* threads to the housekeeping cpus
+else
+    echo "[Manager]" >> /etc/systemd/system.conf
+    echo "CPUAffinity=0 2" >> /etc/systemd/system.conf
+    echo "IRQBALANCE_BANNED_CPUS=FFFFFFFA" >> /etc/sysconfig/irqbalance
```
Comment (vladikr) [inline features/performance/assets/pre-boot-tuning.sh]: These two needs to be calculated as well, based on the isolcpus and reservedSystemCPUs that we set in #23

Comment (yanirq) [reply features/performance/assets/pre-boot-tuning.sh]: we can do that if we set in #23  the reservedCPU value as:
kernelArguments:
  - non_iso_cpu=$reservedCPU 
This means we need to ensure your script runs first

File: features/performance/assets/pre-boot-tuning.sh [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    # Set IRQ banned cpu according to reserved_cpus
+    get_cpu_mask() "0"
+    echo "IRQBALANCE_BANNED_CPUS=$non_iso_cpumask" >> etc/sysconfig/irqbalance
+    
+    find . | cpio -co >../iso_initrd.img
+    cd ..
+    RHCOS_OSTREE_PATH=$(ls -td /boot/ostree/*/ | head -1)
+    cp iso_initrd.img $RHCOS_OSTREE_PATH
+    RHCOS_OSTREE_PATH=${RHCOS_OSTREE_PATH#"/boot"}
+    sed -i "s^initrd .*\$^& ${RHCOS_OSTREE_PATH}iso_initrd.img^" /boot/loader/entries/ostree-2-rhcos.conf2
```
Comment (slintes) [inline features/performance/assets/pre-boot-tuning.sh]: 1. is the  `.conf2` postfix a typo?
2. I don't have an OCP cluster at hand atm, but on the standalone RHCOS I only see `ostree-1-rhcos.conf`. IIRC on OCP you have 2 entries, but how you know which one to modify?

File: features/performance/assets/pre-boot-tuning.sh [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    #TODO - once RHCOS image contains the initrd content we can set parameters with rpm-ostree:
+    #rpm-ostree initramfs --enable --arg=-I --arg=/etc/systemd/system.conf
+    #rpm-ostree initramfs --enable --arg=-I --arg=/etc/sysconfig/irqbalance
+    
+    systemctl reboot
```
Comment (slintes) [inline features/performance/assets/pre-boot-tuning.sh]: talking about reboots... we need to find out how to only trigger 1 reboot when all our configs were applied (at least if we keep single MCs per feature).
Maybe a reboot systemd unit which runs as last unit, checks if we need reboot by searching for a special file, deletes it and then reboots. Just a quick idea.Not needed for this particular PR.

File: features/performance/manifests/01-tuned-rt.yaml [hunk 5]
```diff
@@ -23,10 +23,15 @@ spec:
...
+      [sysfs]
+      /sys/bus/workqueue/devices/writeback/cpumask = ${not_isolated_cpumask}
+      #/sys/devices/virtual/workqueue/cpumask = ${not_isolated_cpumask}
+      /sys/devices/system/machinecheck/machinecheck*/ignore_ce = 1   
+
       [scheduler]
       isolated_cores=${isolated_cores}
     name: openshift-realtime-node
-  recommend:
+  recommend:
```
Comment (slintes) [inline features/performance/manifests/01-tuned-rt.yaml]: does probably not break anything, but looks like we have 2 new trailing spaces here

File: features/performance/assets/pre-boot-tuning.sh (hunk 3 above)
Comment (yanirq) [reply features/performance/assets/pre-boot-tuning.sh]: > I don't have an OCP cluster at hand atm, but on the standalone RHCOS I only see ostree-1-rhcos.conf. IIRC on OCP you have 2 entries, but how you know which one to modify?

I basically followed some given steps to update the boot loader entry file to also load the second initrd which stated to do it ostree-2-rhcos.conf. ill investigate if there is a specific reason for that.

File: features/performance/assets/mc-pre-boot-worker-tuning.yaml.in (hunk 1 above)
Comment (fromanirh) [reply features/performance/assets/mc-pre-boot-worker-tuning.yaml.in]: Agreed, a comment would be most welcome

File: features/performance/manifests/01-tuned-rt.yaml [hunk 6]
```diff
@@ -23,10 +23,15 @@ spec:
       vm.stat_interval = 10
       kernel.timer_migration = 0
 
+      [sysfs]
+      /sys/bus/workqueue/devices/writeback/cpumask = ${not_isolated_cpumask}
+      #/sys/devices/virtual/workqueue/cpumask = ${not_isolated_cpumask}
```
Comment (fromanirh) [inline features/performance/manifests/01-tuned-rt.yaml]: any reason to keep commented code around?

File: features/performance/assets/pre-boot-tuning.sh [hunk 7]
```diff
@@ -2,17 +2,77 @@
...
+    # Set CPU affinity according to reserved_cpus
+    echo "CPUAffinity=$cpu_affinity" >> etc/systemd/system.conf
+
+    # Set IRQ banned cpu according to reserved_cpus
+    get_cpu_mask() "0"
+    echo "IRQBALANCE_BANNED_CPUS=$non_iso_cpumask" >> etc/sysconfig/irqbalance
+    
+    find . | cpio -co >../iso_initrd.img
+    cd ..
+    RHCOS_OSTREE_PATH=$(ls -td /boot/ostree/*/ | head -1)
```
Comment (fromanirh) [inline features/performance/assets/pre-boot-tuning.sh]: I'd add a comment here that we need a more robust approach than keeping the last timestamp.
Until that moment this is good enough I believe.

File: features/performance/manifests/01-tuned-rt.yaml (hunk 6 above)
Comment (yanirq) [reply features/performance/manifests/01-tuned-rt.yaml]: @fromanirh  maybe as a reference of what was done for RHEL but now its done with this patch so ill remove it

File: features/performance/assets/mc-pre-boot-worker-tuning.yaml.in [hunk 8]
```diff
@@ -5,8 +5,8 @@ metadata:
         machineconfiguration.openshift.io/role: worker-rt
     name: 11-rt-kernel-worker-rt
 spec:
-    kernelArguments:
-        - 'non_iso_cpumask=00000005'
+    #kernelArguments:
+    #    - 'non_iso_cpumask=00000005'
```
Comment (fromanirh) [inline features/performance/assets/mc-pre-boot-worker-tuning.yaml.in]: OK, but why?

File: features/performance/assets/pre-boot-tuning.sh [hunk 9]
```diff
@@ -68,7 +68,19 @@ else
...
+    # Get current ostree config file
+    current_ver=1
+    entry_file=""
+    egrep $(uname -r) -lr /boot/loader/entries/ | while read -r line ; do
+        ver=`awk '/version/ {print $2}' $line`
+        if [ "$ver" -gt "$current_ver" ]; then
+           current_ver=$ver
+           entry_file=$line
+        fi
+    done
```
Comment (fromanirh) [inline features/performance/assets/pre-boot-tuning.sh]: nicer indeed, could you please add a couple lines to comment on line 72 to outline the logic?

File: features/performance/manifests/07-mc-pre-boot-worker-tuning.yaml [hunk 10]
```diff
@@ -7,6 +7,7 @@ metadata:
 spec:
     #kernelArguments:
     #    - 'non_iso_cpumask=00000005'
+    #    - 'reserved_cpus'=0-2'
```
Comment (fromanirh) [inline features/performance/manifests/07-mc-pre-boot-worker-tuning.yaml]: why still commented? Is this what we want?

File: features/performance/assets/mc-pre-boot-worker-tuning.yaml.in (hunk 8 above)
Comment (fromanirh) [reply features/performance/assets/mc-pre-boot-worker-tuning.yaml.in]: I mean why we need to comment these lines out now? Please explain in the comments and/or in the commit message

File: features/performance/manifests/07-mc-pre-boot-worker-tuning.yaml [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
...CiAgICBSSENPU19PU1RSRUVfUEFUSD0kKGxzIC10ZCAvYm9vdC9vc3RyZWUvKi8gfCBoZWFkIC0xKQogICAgY3AgaXNvX2luaXRyZC5pbWcgJFJIQ09TX09TVFJFRV9QQVRICiAgICBSSENPU19PU1RSRUVfUEFUSD0ke1JIQ09TX09TVFJFRV9QQVRIIyIvYm9vdCJ9CgogICAgIyBHZXQgY3VycmVudCBvc3RyZWUgY29uZmlnIGZpbGUgYWNjb3JkaW5nIHRvIHRoZSBsYXRlc3QgdmVyc2lvbgogICAgY3VycmVudF92ZXI9MQogICAgZW50cnlfZmlsZT0kKGxzIC10ZCAvYm9vdC9sb2FkZXIvZW50cmllcy8qIHwgaGVhZCAtMSkKICAgIHdoaWxlIHJlYWQgLXIgbGluZSA7IGRvCiAgICAgICAgdmVyPWBhd2sgJy92ZXJzaW9uLyB7cHJpbnQgJDJ9JyAkbGluZWAKICAgICAgICBpZiBbICIkdmVyIiAtZ3QgIiRjdXJyZW50X3ZlciIgXTsgdGhlbgogICAgICAgICAgIGN1cnJlbnRfdmVyPSR2ZXIKICAgICAgICAgICBlbnRyeV9maWxlPSRsaW5lCiAgICAgICAgZmkKICAgIGRvbmUgPDw8JChlZ3JlcCAkKHVuYW1lIC1yKSAtbHIgL2Jvb3QvbG9hZGVyL2VudHJpZXMvKQoKICAgIHNlZCAtaSAic15pbml0cmQgLipcJF4mICR7UkhDT1NfT1NUUkVFX1BBVEh9aXNvX2luaXRyZC5pbWdeIiAkZW50cnlfZmlsZQoKICAgICNUT0RPIC0gb25jZSBSSENPUyBpbWFnZSBjb250YWlucyB0aGUgaW5pdHJkIGNvbnRlbnQgd2UgY2FuIHNldCBwYXJhbWV0ZXJzIHdpdGggcnBtLW9zdHJlZToKICAgICNycG0tb3N0cmVlIGluaXRyYW1mcyAtLWVuYWJsZSAtLWFyZz0tSSAtLWFyZz0vZXRjL3N5c3RlbWQvc3lzdGVtLmNvbmYKICAgICNycG0tb3N0cmVlIGluaXRyYW1mcyAtLWVuYWJsZSAtLWFyZz0tSSAtLWFyZz0vZXRjL3N5c2NvbmZpZy9pcnFiYWxhbmNlCiAgICAKICAgIHN5c3RlbWN0bCByZWJvb3QKZmkK
```
Comment (simon3z) [inline features/performance/manifests/07-mc-pre-boot-worker-tuning.yaml]: @yanirq @MarSik how do we plan to do patch reviews for these blobs?

Comment (yanirq) [reply features/performance/manifests/07-mc-pre-boot-worker-tuning.yaml]: @simon3z  this is a generated manifest,  maybe it should'nt be commit.
The generation should align with @slintes work on https://github.com/openshift-kni/baremetal-deploy/pull/43

Comment (yanirq) [reply features/performance/manifests/07-mc-pre-boot-worker-tuning.yaml]: We can keep it here as a reference / example as in: https://github.com/openshift-kni/baremetal-deploy/blob/22c85c79b537743fbbaa143425244e117749da09/features/rt/manifests/06-mc-rtkernel-worker-rt.yaml#L23

**Pull request merged**